src/kernel/src/obj/copy.rs

```diff
@@ -24,86 +27,489 @@ fn split_range(
+        userinit::create_blank_object,
+    };
+
+    use super::copy_ranges;
+
+    fn copy_ranges_and_check(
+        src: &ObjectRef,
+        src_off: usize,
+        dest: &ObjectRef,
+        dest_off: usize,
+        byte_length: usize,
+    ) {
+        copy_ranges(src, src_off, dest, dest_off, byte_length);
+
+        let dko = kernel_context().insert_kernel_object::<u8>(ObjectContextInfo::new(
+            dest.clone(),
+            Protections::READ,
+            CacheType::WriteBack,
+        ));
+        let dptr = dko.start_addr();
+
+        let sko = kernel_context().insert_kernel_object::<u8>(ObjectContextInfo::new(
+            src.clone(),
+            Protections::READ,
+            CacheType::WriteBack,
+        ));
+        let sptr = sko.start_addr();
+
+        let src_slice = unsafe {
+            core::slice::from_raw_parts(sptr.as_mut_ptr::<u8>().add(src_off), byte_length)
+        };
+        let dest_slice = unsafe {
+            core::slice::from_raw_parts(dptr.as_mut_ptr::<u8>().add(dest_off), byte_length)
+        };
+
+        assert_eq!(src_slice.len(), dest_slice.len());
+        assert!(src_slice == dest_slice);
+    }
+
+    
+    fn zero_ranges_and_check(
+        dest: &ObjectRef,
+        dest_off: usize,
+        byte_length: usize,
+    ) {
+        {
+        let dko = kernel_context().insert_kernel_object::<u8>(ObjectContextInfo::new(
+            dest.clone(),
+            Protections::READ,
+            CacheType::WriteBack,
+        ));
+        let dptr = dko.start_addr();
+        let dest_slice = unsafe {
+            core::slice::from_raw_parts_mut(dptr.as_mut_ptr::<u8>().add(dest_off), byte_length)
+        };
+        dest_slice.fill(0xff);
+        assert!(!dest_slice.iter().all(|x| *x == 0));
         }
+                
+        zero_ranges(dest, dest_off, byte_length);
+
+        let dko = kernel_context().insert_kernel_object::<u8>(ObjectContextInfo::new(
+            dest.clone(),
+            Protections::READ,
+            CacheType::WriteBack,
+        ));
+        let dptr = dko.start_addr();
+        let dest_slice = unsafe {
+            core::slice::from_raw_parts(dptr.as_mut_ptr::<u8>().add(dest_off), byte_length)
+        };
+        assert!(dest_slice.iter().all(|x| *x == 0));
+    }
+
+    #[twizzler_kernel_macros::kernel_test]
+    fn test_object_copy() {
+        let src = create_blank_object();
+        let dest = create_blank_object();
+
+        for p in 0..254u8 {
+            let mut tree: crate::mutex::LockGuard<'_, crate::obj::range::PageRangeTree> =
+                src.lock_page_tree();
+            let (sp, _) =
+                tree.get_or_add_page(PageNumber::base_page().offset(p as usize), true, |_, _| Page::new());
+            sp.as_mut_slice().fill(p + 1);
+        }
+
+        let ps = PageNumber::PAGE_SIZE;
+        let abit = ps / 8;
+        assert!(abit > 0 && abit < ps);
+
+        // Basic test
+        copy_ranges_and_check(&src, ps, &dest, ps, ps);
+
```

Comment:
I'm gonna whine about all the magic numbers below.

Reply:
Will fix.

Reply:
Improvements made. There are still a couple "magic-y" numbers but they are explained and reorganized.